Update OMP

apcraig opened the pull request:
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Update OpenMP implementation
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    The evp1d has changed answers, some unreliable OMP directives were commented out until further debugging can be done.  Tested on several platforms/compiler, Cheyenne, Narwhal, Onyx, Mustang, Gaffney, Cori, Conda, Izumi (which was having system issues), 
- https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#c8b9f17ff8f08536da0e3bc33230f89adc430a22
- https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#d56737c27d4aa0b114b89d1d2dcba97f0319ec59
- https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#0d4c2e39f4308c89a1905b7e05ba2dd723828b2d
- https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#6bb8e22efa8fec5846d5434c519d5274996a9a6e
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit (except evp1d)
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [X] Yes, add omp_suite which tests bit-for-bit with different thread counts
    - [ ] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please provide any additional information or relevant details below:

- Update OpenMP directives as needed including validation via new omp_suite.  Fixed OpenMP in dynamics.  
   - Add SCHEDULE(runtime) to many OMP directive to allow env to set OpenMP schedule.  Add ICE_OMPSCHED to cice.settings to allow the schedule to be user set more easily.  Add set_env test options for a few schedule choices.
   - Add output to log file diagnosing thread schedule on master task
   - Update OMP_STACKSIZE on many platforms to allow OpenMP to run properly.  Default setting was too small causing problems with OpenMP even working
   - VP dynamics still has some debugging to be done (https://github.com/phil-blain/CICE/issues/39, #518), so OpenMP validation was defered
   - A problem was detected in EVP-1d, in ice_dyn_evp_1d_kernel and OpenMP was turned off there.  An issue will be created.
- Refactored eap puny/pi lookups to improve scalar performance
- Update Tsfc implementation to make sure land blocks don't set Tsfc to freezing temp (want it set to zero for consistency across different decompositions/restart files).  Particularly a problem in upwind advection where icepack_compute_tracers is called for every gridpoint (unclear why, but calling it only where tmask is true causes problems).
- Update timers
   - Add ability to use tmp timers, partly commented out.
   - Added timer_stats namelist variable to turn on timer stats
   - Fixed some timer start/stop calls in iblk loops (need to include iblk argument)
   - Add timer updstate (update state calls)
- Ported to Narwhal intel, gnu, cray, aocc
- Add perf_suite (to evaluate CICE performance) and omp_suite (to validate OpenMP)
- Refactored the compare feature in the test scripts and renamed some tests as a result
   - Add ICE_BFBTYPE to cice.settings to define compare process (log, restart, qc, or combinations)
   - Get rid of logbfb test (change to smoke with cmplog option)
   - Get rid of qcchkf and qcnonbfb (use qcchk with qcchkf option)
   - Add several new test options
- Update documentation

apcraig commented:
Closes #114.

dabail10 reviewed (approved):
I am impressed by the amount of work here. I can't see any obvious problems here. I am a bit concerned about the Tsfc values on land. However, I agree with what you have done.

apcraig commented:
[CICE_OMP_perf_220119.xlsx](https://github.com/CICE-Consortium/CICE/files/7901243/CICE_OMP_perf_220119.xlsx)

CICE performance on several machines is summarized in the attached Excel spreadsheet for this version of the code.  In particular, the three plots show Dynamics, Advection, and Column times for a fixed problem (gx1, 64 cores, 16x16 block size) varying tasks and threads (always running on 64 cores).  OpenMP is working reasonably OK.  It's still not generally faster than all MPI (but is in a few cases and is close in many others).  This also shows the relative difference across a few machines and compilers.  Things are working better now than they were before, especially now that dynamics OMP is turned on.

The other thing to note in the attached spreadsheet is the difference in time between (scheds, scheds1, schedd1).  These are different OMP schedules.  scheds="STATIC", scheds1="STATIC,1", schedd1="DYNAMIC,1".  You can see cases where the dynamic scheduling overhead is not good (onyx_intel) and other cases where the dynamic scheduling does relatively better.  In general, the scheduling has a greater impact on advection and column timers probably because load balance across work is more important in those parts of the code.

A couple other notes.  64x0 is 64 MPI tasks with threading turned off in compilation.  64x1 is 64 MPI tasks with 1 thread per task compiled with threading on.

phil-blain commented:
Just a note that I'm on vacation this coming week, but I'd like to take a look at this when I'm back.

eclare108213 commented on cicecore/cicedynB/infrastructure/comm/mpi/ice_timers.F90:
```diff
@@ -62,8 +65,18 @@ module ice_timers
+!      timer_tmp2,             &! for temporary timings
+!      timer_tmp3,             &! for temporary timings
+!      timer_tmp4,             &! for temporary timings
+!      timer_tmp5,             &! for temporary timings
```
are you planning to clean these tmp timers up later?

apcraig replied:
I go back and forth on this.  I was thinking I'd just leave them there but commented out.  I find when I'm debugging any performance thing, I end up implementing the same temporary timers and dropping them in where I need a finer detail look.  Then removing them at the end.  It might be nice to just have them readily available anytime anyone needs to use them.  I can remove them if that's preferred.

eclare108213 replied:
I'm ambivalent.  Sometimes it's nice to have commented-out code sitting around for ease of use.  Comments/opinions from others are welcome!

eclare108213 commented on cicecore/cicedynB/infrastructure/ice_grid.F90:
```diff
@@ -406,6 +414,29 @@ subroutine init_grid2
+      ! Diagnose OpenMP thread schedule, force order in output
+      !-----------------------------------------------------------------
+
+#if defined (_OPENMP)
```
why is this done in ice_grid.F90?

apcraig replied:
It's a good question.  It has to be done after the decomposition is established.  There is no obvious place for something like this in the code, but I'm happy to move it elsewhere.  I wanted to keep it out of the driver code because then it would need to be duplicated in all of them.  And it needed to be somewhere that is always called, so it couldn't be a "first call" thing in the column physics or dynamics.  I am open to moving this to another part of the code.  I think it's handy to have this information diagnosed, there is no way to tell otherwise.  In fact, I added this stuff because I wanted to verify the OMP schedule was working.  That's how I learned it wasn't and how we need to add the SCHEDULE(runtype) argument to the OMP pragma.  This is very confusing.  I pointed the problems out to the cheyenne help desk since they were helping me with some of this.  I proposed a change to the OpenMP standards to fix this and preserve backwards compatibility, but don't know if it'll go anywhere.

eclare108213 replied:
I was thinking that it belonged in the module where decompositions are done (ice_blocks?).  But I guess that necessitates calling it from every driver.

Interesting that you proposed a change to the OpenMP standards!  We should track that in Consortium reporting docs.

phil-blain replied:
Do you have a link for your proposition @apcraig ? I'd like to follow any developments:)

apcraig replied:
I forwarded my input to NCAR who has a person on the OpenMP committee.  Nothing more formal than that.

eclare108213 commented on configuration/scripts/machines/env.narwhal_aocc:
```diff
@@ -33,6 +33,8 @@ module load cray-hdf5/1.12.0.4
 setenv NETCDF_PATH ${NETCDF_DIR}
 limit coredumpsize unlimited
 limit stacksize unlimited
+setenv OMP_STACKSIZE 128M
```
how do you know what to set OMP_STACKSIZE to? Do you provide guidance on this in the user docs?

apcraig replied:
I did add that information to the documentation (see https://github.com/CICE-Consortium/CICE/pull/680/files#diff-ea684240af50d1d378a32f8805b71db3773d3eb5df8342f1bf262637bc929529).  In fact, that was one of the things that prevented the OpenMP from working on several platforms.  Many platforms had the default set to 0 or 4M which caused seg faults in OpenMP.  When I increased the value, things worked.  There wasn't always a problem in the OpenMP implementation in CICE that caused failures.  Cheyenne has the default set to 64M which is why things were often working there but not elsewhere and gave me a clue.  This took a while to figure out.  I don't know what value is the minimum needed for CICE.  It will depend on the problem size, number of tasks and threads, decomposition, etc.  I'm pretty sure the tests I was running will work with 16M.  So I set the defaults higher than that on some of the platforms where the default was too small.

eclare108213 replied:
Is it set to at least 16M for all currently supported machines?   Why are some machines as high as 128M, if 16M is enough for the current test suites?

apcraig replied:
I only changed the setting for the machines I have access to and tested.  I'm reluctant to set values for machines I don't have access to.  They may already have reasonable defaults and/or settings that the systems folks prefer you use.  I recognize that I set some machines to 32M, 64M, and 128M.  That was not a careful process.  For other machines, I just left the default value because it worked.  I tried different values on different machines to see what would work and if it worked, I just stuck with it.  I did not try to figure out the minimum value needed for testing or by machine/compiler.  The compilers/implementation may also change in the future, so it makes sense to set it above the minimum required value anyway.  All the values are pretty reasonable as is.   If the OMP_STACKSIZE needed to be set so high that if affected other parts of the memory use, then we'd want to be more careful.  As it is, I think we have a pretty large range that OMP_STACKSIZE could be set without affecting other aspects of the memory/performance (say 16M to 256M) for our testing.  I don't think we need to be particularly precise here.

eclare108213 reviewed (commented):
Wow, I'm impressed.  This looks like an enormous amount of work!  I'd really like to know what your workflow was for identifying OMP problems and fixing them.

Just a few questions, including this one:  you refactored the compare feature in the test scripts and renamed some tests as a result -- maybe I missed it, but I didn't see relevant changes to the documentation.  Are your mods, e.g. for running QC tests, invisible for users?

TillRasmussen commented:
@apcraig This looks really nice. Approved from here.

apcraig commented:
I will review the qc documentation.  You're right that I probably need to make a few minor updates there.

With regard to workflow, it was a bit of everything.  I was planning to use debugging/performance tools if needed.  But I started with just testing things carefully.  I added some new test suites to check the OpenMP results.  I reviewed each OMP pragma line carefully (for private variables) and turned them on and off to check we were getting identical results.  I found problems (like the timer iblk implementation) as I went and fixed those things, that helped quite a bit.  I tested on multiple platforms and compilers which is a big pain.  But that was a key.  I used "setenv OMP_DISPLAY_ENV TRUE" to write out the OMP settings at run time.  This helped me pinpoint differences in defaults for machines where the OpenMP was working and machines where it wasn't.  There was a lot of trial and error and confusion.  I spent a good part of 3 weeks on this, so it took a bit of time.  It was good I had a window to focus.  I don't think I could have made the same progress working part-time on it.  Lots of googling too.  In the end, I didn't need the debugging and performance tools.  Our timers were working well.  I also didn't know how I would detect missing private OMP variables via a debugger.  I guess you track where things diverge and then start digging there, but understanding that the problem was partly the default OpenMP setting was an important piece why we were getting inconsistent results on different machines and compilers.

apcraig commented:
I reviewed the qc documentation and it's fine.  It's only the automated testing via test suites that is changed.  The manual process is not.

apcraig commented:
@dabail10 regarding the Tsfc values on land.  This was a problem with testing, not results.  The exact restart comparisons were failing in some cases with different decomps/tasks/threads because Tsfc was being set to the freezing temperature on land in some cases.  With different decomps/tasks/threads, the land block elimination changes, and eliminated land blocks had values of zero for Tsfc but freezing temperature for land that was not eliminated making the restarts not identical across different test cases.  It turns out, this was caused primarily by the upwind advection scheme and showed up in just a few tests.  I tried to simply stop computation of the upwind advection on the land cells, but that changed the answers (which is scary, but I decided not to dig deeper).  So I needed to mask the Tsfc in the upwind advection after the computation was complete.  That's the change in line 1745 of ice_dyn_transport.F90.  The other changes are not necessarily required, but I added them for good measure.  There could be cases where a restart file has non-zero Tsfc on the land in the restart file and then values are never reset to zero.  Anyway, the eliminated land always has values of zero when restart/history files are written.  For the most part, all fields in CICE on land are also zero, so that's fine.  Tsfc is one field where this wasn't entirely the case.

rgrumbine commented:
As a matter of taste, I prefer to keep debugging code sitting around for
the next round of debugging. One thing I do, though, is to make that part
of the line, e.g. rather than
!    timer_tmp,   &! for temporary timing

I do something like
!debug     timer_tmp,   &! for temporary timing



On Mon, Jan 24, 2022 at 3:12 PM Elizabeth Hunke ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In cicecore/cicedynB/infrastructure/comm/mpi/ice_timers.F90
> <https://github.com/CICE-Consortium/CICE/pull/680#discussion_r791115057>:
>
> > @@ -62,8 +65,18 @@ module ice_timers
>        timer_bgc,              &! biogeochemistry
>        timer_forcing,          &! forcing
>        timer_evp_1d,           &! timer only loop
> -      timer_evp_2d             ! timer including conversion 1d/2d
> -!      timer_tmp               ! for temporary timings
> +      timer_evp_2d,           &! timer including conversion 1d/2d
> +      timer_updstate           ! update state
> +!      timer_updstate,         &! update state
> +!      timer_tmp1,             &! for temporary timings
> +!      timer_tmp2,             &! for temporary timings
> +!      timer_tmp3,             &! for temporary timings
> +!      timer_tmp4,             &! for temporary timings
> +!      timer_tmp5,             &! for temporary timings
>
> I'm ambivalent. Sometimes it's nice to have commented-out code sitting
> around for ease of use. Comments/opinions from others are welcome!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CICE-Consortium/CICE/pull/680#discussion_r791115057>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFQ65HNMVT46ZE73XECH553UXWXC7ANCNFSM5MKXG55A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Tel: 301-683-3747
Coupling and Dynamics Group
(he/him/his)

TillRasmussen commented:
The issue with the OMP within the 1d has been tracked. See #681.

eclare108213 commented:
@phil-blain just pinging you again on this one, when you have a chance.  thx!

apcraig commented:
It would be good to merge this today or tomorrow if we can.....  Would like to get the weekend testing operating on these updates as well as possible merge to the C/CD branch.  Thanks.

phil-blain commented:
Sorry, I'll review it first thing tomorrow morning.

phil-blain commented on cicecore/cicedynB/dynamics/ice_transport_remap.F90:
```diff
@@ -448,17 +453,34 @@ subroutine horizontal_remap (dt,                ntrace,     &
-      !$TCXOMP          indxinc,indxjnc,mmask,tmask,istop,jstop,l_stop)
+      !$OMP PARALLEL DO PRIVATE(iblk,ilo,ihi,jlo,jhi,this_block,n,   &
+      !$OMP          indxinc,indxjnc,mmask,tmask,istop,jstop,l_stop) &
+      !$OMP SCHEDULE(runtime)
```
Does it still lead to a segfault with GCC  ? if not the comment could be removed.

apcraig replied:
Those comment lines are already removed.  Wonder why you're not seeing that.  Check the current file diffs again.

phil-blain replied:
Yeah, I realized that after submitting my review, It's because I'm reviewing commit by commit... sorry for the noise.

phil-blain commented on cicecore/cicedynB/dynamics/ice_transport_remap.F90:
```diff
@@ -632,12 +679,21 @@ subroutine horizontal_remap (dt,                ntrace,     &
+      !$OMP                     edgearea_e,edgearea_n,edge,iflux,jflux, &
+      !$OMP                     xp,yp,indxing,indxjng,mflxe,mflxn, &
+      !$OMP                     mtflxe,mtflxn,triarea,istop,jstop,l_stop) &
+      !$OMP SCHEDULE(runtime)
```
same here, if it works.

apcraig replied:
Same here.  I check both the file diffs in the PR and the code in my sandbox.

phil-blain commented on cicecore/drivers/standalone/cice/CICE_RunMod.F90:
```diff
@@ -160,7 +161,11 @@ subroutine ice_step
+          timer_tmp1, timer_tmp2, timer_tmp3, timer_tmp4, timer_tmp5, timer_tmp6
+#ifdef OMP_TIMERS
+      use OMP_LIB
+#endif
```
does it work with ` omp_lib` (I'm thinking it should) ? that would make case more consistent with our own modules.

apcraig replied:
This code does not exist in this version.   I'm confused what's happening here.

phil-blain commented on cicecore/drivers/standalone/cice/CICE_RunMod.F90:
```diff
@@ -318,6 +383,9 @@ subroutine ice_step
 
+#ifdef OMP_TIMERS
+         call ice_barrier()
+#endif
```
I'm curous why we need a barrier at a few places ? this is an MPI barrier, no? How does it interact with OpenMP ?

apcraig replied:
Same.

dabail10 commented:
Just an update here. I am finding some divide by zero issues in icepack. I am puzzled why this is only a problem when doing a restart with CD/C grids. I assume there is some mask issue where icepack is being called for a non-ice point?

phil-blain commented on cicecore/drivers/standalone/cice/CICE_RunMod.F90:
```diff
@@ -386,6 +480,7 @@ subroutine ice_step
          call accum_hist (dt)               ! history file
          call ice_timer_stop(timer_hist)    ! history
 
+         call ice_barrier()
```
and this one is not protected by `OPM_TIMERS`. is that what we want ?

apcraig replied:
Same.

phil-blain commented on configuration/scripts/cice.run.setup.csh:
```diff
@@ -43,7 +41,9 @@ set ICE_RUNLOG_FILE = "cice.runlog.\${stamp}"
 cd \${ICE_RUNDIR}
 
-setenv OMP_NUM_THREADS ${nthrds}
+setenv OMP_NUM_THREADS \${ICE_NTHRDS}
```
On some systems, it's only via the job scheduler that you can specify the number of OpenMP threads, changing `OMP_NUM_THREADS` at runtime has no effect. I think it's the case at least for PBS Pro, the number of threads is set from the `ompthreads` directive in the select statement. Maybe that would be worth to mention in the documentation ? (maybe not, just thinking out loud...)

apcraig replied:
I wasn't aware of that.  I guess that's part of setting up the port properly.

eclare108213 commented:
@dabail10 I'm confused - is this question in the wrong place (OMP PR)?

phil-blain commented on doc/source/user_guide/ug_implementation.rst:
```diff
@@ -1083,19 +1085,19 @@ The timers use *MPI\_WTIME* for parallel runs and the F90 intrinsic
-   | 8            | Ridging     | mechanical redistribution                          |
+   | 8            | Ridging     | mechanical redistribution, part of Column timer    |
    +--------------+-------------+----------------------------------------------------+
-   | 9            | FloeSize    | flow size                                          |
```
nice additions!

dabail10 commented:
Whoops.

phil-blain reviewed (commented):
I had a thorough look at most of the commits, and skimmed through the following ones:
- [Update EVP OMP implementation](https://github.com/CICE-Consortium/CICE/pull/680/commits/3446cf632793432536dc14f080d6d80c5df4bf5c)

- [Refactor puny/pi scalars in eap dynamics to improve performance](https://github.com/CICE-Consortium/CICE/pull/680/commits/f91f4597b8d0740227a6172b9b496a2f1976ac29)

- [Clean up](https://github.com/CICE-Consortium/CICE/pull/680/commits/7905bf69d444a84e9f111e2a75543fb49058c526)


- [Comment out temporary timers](https://github.com/CICE-Consortium/CICE/pull/680/commits/667e998c9ac61bf66eb781659e0385fd693aecf9)

It mostly looks good, thanks a lot for this work. It's still on my list to go back and fix the noted issues in the VP module, including OpenMP validation. This here PR should be a good start.
